graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -584,6 +628,305 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+    return visitor_fn
+
+
+def _get_present_locations_from_match_traversals(match_traversals):
+    """Return the set of locations present in the given match traversals.
+
+    When enumerating the possibilities for optional traversals,
+    the resulting match traversals may have sections of the query omitted.
+    These locations will not be included in the returned `present_locations`.
+
+    Args:
+        match_traversals: one possible list of match traversals generated from a query
+            containing @optional traversal(s)
+
+    Returns:
+        present_locations: set of all locations present in the given match traversals
+    """
+    present_locations = set()
+
+    for match_traversal in match_traversals:
+        for step in match_traversal:
+            if step.as_block is not None:
+                location_name, _ = step.as_block.location.get_location_name()
+                present_locations.add(location_name)
+    return present_locations
+
+
+def _lower_filters_in_match_traversals(match_traversals, visitor_fn):
+    """Return new match traversals, lowering filters involving non-existent ContextFields.
+
+    Expressions involving non-existent ContextFields are evaluated to True.
+    BinaryCompositions, where one of the operands is lowered to a TrueLiteral,
+    are lowered appropriately based on the present operator (u'||' and u'&&' are affected).
+    TernaryConditionals, where the predicate is lowerd to a TrueLiteral,
+    are replaced by their if_true predicate.
+    The `visitor_fn` implements these behaviors (see `_update_context_field_expression`).
+
+    Args:
+        match_traversals: list of match traversal enities to be lowered
+        visitor_fn: visit_and_update function for lowering expressions in given match traversal
+
+    Returns:
+        new_match_traversals: new list of match_traversals, with all filter expressions lowered
+    """
+    new_match_traversals = []
+    for match_traversal in match_traversals:
+        new_match_traversal = []
+        for step in match_traversal:
+            if step.where_block is not None:
+                new_filter = step.where_block.visit_and_update_expressions(visitor_fn)
+                if new_filter.predicate == TrueLiteral:
+                    new_filter = None
+                new_step = step._replace(where_block=new_filter)
+            else:
+                new_step = step
+            new_match_traversal.append(new_step)
+        new_match_traversals.append(new_match_traversal)
+    return new_match_traversals
+
+
+def lower_context_field_expressions_in_compound_match_query(compound_match_query):
+    """Lower Expressons involving non-existent ContextFields."""
+    if len(compound_match_query.match_queries) == 0:
+        raise AssertionError(u'Received CompoundMatchQuery with an empty list of MatchQueries.')
+    elif len(compound_match_query.match_queries) == 1:
+        # All ContextFields exist if there is only one MatchQuery.
```

Comment:
Elaborate? I think I know why, but I don't think I'll remember my reasoning in six months.